### DISCUSSION TITLE
autogenerate_proxy_classes and proxies understanding

### DISCUSSION THREAD

**Discussion by FrancescJR** [doctrine/doctrine2#10386] [General]
Hello,

I just would like to understand more about the autogenerate_proxy_classes configuration. Maybe I don't need to understand it all, but just understand a specific behavior I am having that contradicts my understanding, hence the need for help.

So what I understand:
1. doctrine uses proxy objects (explained here: https://www.doctrine-project.org/projects/doctrine-orm/en/2.11/reference/advanced-configuration.html#auto-generating-proxy-classes-optional)
2. Proxy classes can either be generated manually or automatically at runtime by Doctrine  (extract from docs).
I am not sure if I have a configuration with this `auto_generate_proxy_classes: true` if that means it's automatically, I'd say so, right?

Now, what I don't understand is if proxy classes are eventually always used (independently if the config option) or just in some cases. I'd say that they are eventually always needed, despite on how they're generated, right?

But the behavior I'm having is:
if `auto_generate_proxy_classes: true` everything works fine (I can save entites to database normally)
if `auto_generate_proxy_classes: false` then I get this error: https://github.com/doctrine/orm/issues/7598  (yes the entity is "final")

But I'd say I should get the error always, maybe sooner or later, but I'd get the error too if I have a final class even though I have this `auto_generate_proxy_classes: true`, right?

Can somebody explain me the why behind all of this?

**Comment by MatTheCat** [comment]
Three years later…

Every entity doesn’t require a proxy: e.g. one which no relationship has as a target. If proxies are runtime-generated then it wouldn’t ever happen for such an entity.

But the code generating proxies if `auto_generate_proxy_classes` is `false` doesn’t know that, so it will always generate proxies for every entity. As such it can make this kind of issue appear.

**Reply by FrancescJR** [reply]
Thanks! Even if it's after 3 years.
